src/contracts/services/silkBasket.ts

```diff
@@ -0,0 +1,328 @@
+    // calculated values. Query will need to be retried with modified oracle keys.
+    if (!matchingPriceResponse) {
+      status = SilkBasketParsingStatus.PRICES_MISSING;
+      return {
+        symbol: item.symbol,
+        amount: item.weight.fixed,
+        price: undefined,
+        weight: {
+          initial: item.weight.initial,
+          current: undefined,
+        },
+      };
+    }
+
+    // confirmed price is available so we will return the price/weight calculated data
+    const itemPrice = convertCoinFromUDenom(
+      matchingPriceResponse.data.rate,
+      ORACLE_NORMALIZATION_FACTOR,
+    );
+    const currentWeight = itemPrice.multipliedBy(item.weight.fixed).dividedBy(silkPrice);
+    return {
+      symbol: item.symbol,
+      amount: item.weight.fixed,
+      price: itemPrice.toString(),
+      weight: {
+        initial: item.weight.initial,
+        current: currentWeight.toString(),
+      },
+    };
+  });
+
+  return {
+    symbol: silkBasketResponse.symbol,
+    oracleRouterContract: {
+      address: silkBasketResponse.router.address,
+      codeHash: silkBasketResponse.router.code_hash,
+    },
+    staleInterval: Number(silkBasketResponse.when_stale),
+    peg: {
+      value: silkPrice.toString(),
+      initialValue: silkInitialPrice.toString(),
+      isFrozen: silkBasketResponse.peg.frozen,
+      lastUpdatedAt: new Date(Number(silkBasketResponse.peg.last_updated) * 1000),
+    },
+    basket: basketItems,
+    blockHeight: silkBasketResponseBlockHeight,
+    status,
+  };
+};
+
+/**
+ * parses the silk basket and prices response via the query router
+ */
+function parseSilkBasketAndPricesResponseFromQueryRouter(
+  batchQueryResponse: BatchQueryParsedResponse,
+) {
+  const responseCount = batchQueryResponse.length;
+  if (responseCount !== 2) {
+    throw new Error(`${responseCount} responses found, two responses are expected`);
+  }
+
+  const silkBasketResponse = batchQueryResponse.find((
+    responseItem,
+  ) => responseItem.id === SilkBasketBatchResponseItem.BASKET);
+
+  // check if data exists
+  if (!silkBasketResponse) {
+    throw new Error('Silk basket response not found in query router response');
+  }
+  // handle error state
+  if (silkBasketResponse.status === BatchItemResponseStatus.ERROR) {
+    throw new Error(silkBasketResponse.response);
+  }
+
+  const pricesResponse = batchQueryResponse.find((
+    responseItem,
+  ) => responseItem.id === SilkBasketBatchResponseItem.PRICES);
+
+  // check if data exists
+  if (!pricesResponse) {
+    throw new Error('prices response not found in query router response');
+  }
+
+  // for price error state we will not throw an error here as there is still useful
+  // data to get out of the peg info that we would like to parse.
+  // Instead, we will pass the full batch data into the parser (which includes the
+  // error status) and handle the error downstream.
+
+  return parseSilkBasketAndPricesResponse({
```

Comment:
i could possible make one big parser and not call this function. The separation made more sense in scenarios when working with a single query wrapped in a batch query (for block height purposes).